rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/rename/Common.rsc

```diff
@@ -0,0 +1,224 @@
+
+set[Define] rascalReachableDefs(TModel tm, set[loc] defs) {
+    rel[loc from, loc to] modulePaths = rascalGetTransitiveReflexiveModulePaths(tm);
+    rel[loc from, loc to] scopes = rascalGetTransitiveReflexiveScopes(tm);
+    rel[loc from, Define define] reachableDefs =
+        ((tm.defines<defined, defined, scope>)[defs]             // <definition, scope> pairs
+        + (tm.defines<scope, defined, scope>)[defs])
+      o (
+         (scopes                                                // All scopes surrounding defs
+        o modulePaths                                           // Transitive-reflexive paths from scope to reachable modules
+        ) + (                                                   // In ADT and syntax definitions, search inwards to find field scopes
+            (tm.defines<idRole, scope, defined>)[{dataId(), nonterminalId(), lexicalId()}]
+          + tm.defines<scope, scope>
+        )
+      )
+      o tm.defines<scope, defined>                              // Definitions in these scopes
+      o definitionsRel(tm);                                     // Full define tuples
+    return reachableDefs.define;                                // We are only interested in reached defines; not *from where* they were reached
+}
+
+@memo{maximumSize(1), expireAfter(minutes=5)}
+rel[loc from, loc to] rascalGetTransitiveReflexiveModulePaths(TModel tm) {
+    rel[loc from, loc to] moduleI = ident(getModuleScopes(tm));
+    rel[loc from, loc to] imports = (tm.paths<pathRole, from, to>)[importPath()];
+    rel[loc from, loc to] extends = (tm.paths<pathRole, from, to>)[extendPath()];
+
+    return (moduleI + imports)  // 0 or 1 imports
+         o (moduleI + extends+) // 0 or more extends
+         ;
+}
+
+set[loc] rascalGetOverloadedDefs(TModel tm, set[loc] defs) {
+    if (defs == {}) return {};
+
+    set[Define] overloadedDefs = {tm.definitions[d] | d <- defs};
+    set[IdRole] roles = overloadedDefs.idRole;
+
+    // Pre-conditions
+    assert size(roles) == 1:
+        "Initial defs are of different roles!";
+    assert rascalMayOverloadSameName(defs, tm.definitions):
+        "Initial defs are invalid overloads!";
+
+    IdRole role = getFirstFrom(roles);
+    map[loc file, loc scope] moduleScopePerFile = getModuleScopePerFile(tm);
+    rel[loc def, loc scope] defUseScopes = {<d, moduleScopePerFile[u.top]> | <loc u, loc d> <- tm.useDef};
+    rel[loc fromScope, loc toScope] modulePaths = rascalGetTransitiveReflexiveModulePaths(tm);
+
+    rel[loc def, loc moduleScope] defPathStep =
+        (tm.defines<defined, scope>+ + defUseScopes) // 1. Look up scopes of defs and scopes of their uses
+        o (modulePaths + invert(modulePaths))        // 2. Follow import/extend relations to reachable scopes
+        ;
+
+    rel[loc fromDef, loc toDef] defPaths = {};
+    set[loc] reachableDefs = rascalReachableDefs(tm, overloadedDefs.defined).defined;
+
+    solve(overloadedDefs) {
+        if (constructorId() := role) {
```

Comment:
nit: this long chain of `if`s could also be written as a `switch`.

Maybe come of the bodies could be moved to separate functions?

Reply:
Code removed